File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/library/items/list/DemoableItemListAdapter.java
package com.lasthopesoftware.bluewater.client.library.items.list;

import android.app.Activity;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.view.View;
import android.view.ViewGroup;
import com.lasthopesoftware.bluewater.R;
import com.lasthopesoftware.bluewater.client.library.items.IItem;
import com.lasthopesoftware.bluewater.client.library.items.list.menus.changes.handlers.IItemListMenuChangeHandler;
import com.lasthopesoftware.bluewater.client.library.items.media.files.access.parameters.IFileListParameterProvider;
import com.lasthopesoftware.bluewater.client.library.repository.Library;
import com.lasthopesoftware.bluewater.client.stored.library.items.StoredItemAccess;
import com.lasthopesoftware.bluewater.shared.MagicPropertyBuilder;
import tourguide.tourguide.Overlay;
import tourguide.tourguide.Pointer;
import tourguide.tourguide.ToolTip;
import tourguide.tourguide.TourGuide;

import java.util.List;

public class DemoableItemListAdapter<T extends IItem> extends ItemListAdapter<T> {

	private static final String PREFS_KEY = MagicPropertyBuilder.buildMagicPropertyName(DemoableItemListAdapter.class, "TUTORIAL_SHOWN");
	private static final boolean DEBUGGING_TUTORIAL = false;
	private final Activity activity;
	private final List<T> items;

	private boolean wasTutorialShown;

	public DemoableItemListAdapter(Activity activity, int resource, List<T> items, IFileListParameterProvider<T> fileListParameterProvider, IItemListMenuChangeHandler itemListMenuEvents, StoredItemAccess storedItemAccess, Library library) {
		super(activity, resource, items, fileListParameterProvider, itemListMenuEvents, storedItemAccess, library);

		this.activity = activity;
		this.items = items;
	}

	@NonNull
	@Override
	public View getView(int position, View convertView, @NonNull ViewGroup parent) {
		final View view = super.getView(position, convertView, parent);
		if (items.isEmpty() || (items.size() > 1 && position != 2)) return view;

		buildTutorialView(view);
		return view;
	}

	private void buildTutorialView(final View view) {
		// use this flag to ensure the least amount of possible work is done for this tutorial
		if (wasTutorialShown) return;
		wasTutorialShown = true;

		final SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(activity);
		if (!DEBUGGING_TUTORIAL && sharedPreferences.getBoolean(PREFS_KEY, false)) return;

		final int displayColor = activity.getResources().getColor(R.color.clearstream_blue);

		final TourGuide tourGuide =
			TourGuide.init(activity).with(TourGuide.Technique.CLICK)
				.setPointer(new Pointer().setColor(displayColor))
				.setToolTip(new ToolTip()
					.setTitle(activity.getString(R.string.title_long_click_menu))
					.setDescription(activity.getString(R.string.tutorial_long_click_menu))
					.setBackgroundColor(displayColor))
				.setOverlay(new Overlay())
				.playOn(view);

		view.setOnLongClickListener(v -> {
			tourGuide.cleanUp();
			view.setOnLongClickListener(null);
			return false;
		});

		sharedPreferences.edit().putBoolean(PREFS_KEY, true).apply();
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/browsing/items/list/DemoableItemListAdapter.kt
package com.lasthopesoftware.bluewater.client.browsing.items.list

import android.app.Activity
import android.os.Build
import android.view.View
import com.lasthopesoftware.bluewater.R
import com.lasthopesoftware.bluewater.client.browsing.items.access.ProvideItems
import com.lasthopesoftware.bluewater.client.browsing.items.list.menus.changes.handlers.IItemListMenuChangeHandler
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.access.parameters.IFileListParameterProvider
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.access.stringlist.FileStringListProvider
import com.lasthopesoftware.bluewater.client.browsing.library.repository.Library
import com.lasthopesoftware.bluewater.client.stored.library.items.StoredItemAccess
import com.lasthopesoftware.bluewater.shared.android.messages.SendMessages
import com.lasthopesoftware.bluewater.shared.promises.extensions.LoopedInPromise
import com.lasthopesoftware.bluewater.tutorials.ManageTutorials
import com.lasthopesoftware.bluewater.tutorials.TutorialManager
import tourguide.tourguide.Overlay
import tourguide.tourguide.Pointer
import tourguide.tourguide.ToolTip
import tourguide.tourguide.TourGuide

class DemoableItemListAdapter
(
	private val activity: Activity,
	private val sendMessages: SendMessages,
	fileListParameterProvider: IFileListParameterProvider,
	fileStringListProvider: FileStringListProvider,
	itemListMenuEvents: IItemListMenuChangeHandler,
	storedItemAccess: StoredItemAccess,
	provideItems: ProvideItems,
	library: Library,
	private val manageTutorials: ManageTutorials
) : ItemListAdapter(
	activity,
	sendMessages,
	fileListParameterProvider,
	fileStringListProvider,
	itemListMenuEvents,
	storedItemAccess,
	provideItems,
	library
) {
	private var wasTutorialShown = false

	override fun onBindViewHolder(holder: ViewHolder, position: Int) {
		super.onBindViewHolder(holder, position)
		if (itemCount == 1 || position != 2)
			bindTutorialView(holder.itemView)
	}

	private fun bindTutorialView(view: View) {
		// use this flag to ensure the least amount of possible work is done for this tutorial
		if (wasTutorialShown) return

		wasTutorialShown = true

		fun showTutorial() {
			val displayColor =
				if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)	activity.resources.getColor(R.color.clearstream_blue, null)
				else activity.resources.getColor(R.color.clearstream_blue)

			val tourGuide = TourGuide.init(activity).with(TourGuide.Technique.CLICK)
				.setPointer(Pointer().setColor(displayColor))
				.setToolTip(
					ToolTip()
						.setTitle(activity.getString(R.string.title_long_click_menu))
						.setDescription(activity.getString(R.string.tutorial_long_click_menu))
						.setBackgroundColor(displayColor)
				)
				.setOverlay(Overlay())
				.playOn(view)

			view.setOnLongClickListener {
				tourGuide.cleanUp()
				view.setOnLongClickListener(null)
				false
			}
		}

		if (DEBUGGING_TUTORIAL) {
			showTutorial()
			return
		}

		manageTutorials
			.promiseWasTutorialShown(TutorialManager.KnownTutorials.longPressListTutorial)
			.eventually(LoopedInPromise.response({ wasShown ->
				if (!wasShown) {
					showTutorial()
					manageTutorials.promiseTutorialMarked(TutorialManager.KnownTutorials.longPressListTutorial)
				}
			}, activity))
	}

	companion object {
		private const val DEBUGGING_TUTORIAL = false
	}
}
